### PR TITLE
Improve documentation and remove a few private things from the public API

### PR DESCRIPTION
**Pull request by nickpape-msft**
* Removed quite a few of the logging utilities. We may want to expose these again in the future, but if we do, we should remove the helper functions on the GulpTask.

* **State** is completely messed up in GCB. IMO we should do a PR which will refactor all state to come from a single place. There are too many places we pass around references to the `buildConfig` object, and I know from experience that there are many places we are not even using the `./config.ts` file!

* Added lots of comments.

* One tiny refactoring that removes an unnecessary function.

### CODE REVIEW

**Comment by nickpape-msft** [comment]
@dzearing Are you using any of the removed API's ???

File: gulp-core-build/src/tasks/GulpTask.ts [hunk 1]
```diff
@@ -83,40 +110,98 @@ export abstract class GulpTask<TASK_CONFIG> implements IExecutable {
...
+   * @param buildConfig - the build configuration which should be used when determining if the task is enabled
+   * @returns true if the build is not redundant
+   */
   public isEnabled(buildConfig: IBuildConfig): boolean {
     return (!buildConfig || !buildConfig.isRedundantBuild);
   }
 
+  /**
+   * When the task is executed by the build system, this function is called once. Note that this function
+   * must either return a Promise, a Stream, or call the completeCallback() parameter.
```
Comment (iclanton) [inline gulp-core-build/src/tasks/GulpTask.ts]: I believe it must always call the `completeCallback` callback.

File: gulp-core-build/src/tasks/GulpTask.ts [hunk 2]
```diff
@@ -83,40 +110,98 @@ export abstract class GulpTask<TASK_CONFIG> implements IExecutable {
...
+   * by the CleanTask.
+   * @param buildConfig - the current build configuration
+   * @param taskConfig - a task instance's configuration
+   */
   public getCleanMatch(buildConfig: IBuildConfig, taskConfig: TASK_CONFIG = this.taskConfig): string[] {
     return this.cleanMatch;
   }
 
+  /**
+   * This function is once to execute the task. It calls executeTask() and handles the return
```
Comment (iclanton) [inline gulp-core-build/src/tasks/GulpTask.ts]: Looks like you a word.

File: gulp-core-build/src/tasks/ValidateShrinkwrapTask.ts [hunk 3]
```diff
@@ -6,11 +6,17 @@ import * as semver from 'semver';
 type ShrinkwrapDep = { [name: string]: { version: string } };
 type PackageDep = { [name: string]: string };
 
+/**
+ * Partial representation of the contents of a `package.json` file
```
Comment (iclanton) [inline gulp-core-build/src/tasks/ValidateShrinkwrapTask.ts]: What does "Partial" mean in these two contexts?

Comment (nickpape-msft) [inline gulp-core-build/src/tasks/ValidateShrinkwrapTask.ts]: Partial, meaning that the this is not an interface to represent the entire package.json file.

File: common/changes/nickpape-improve-api_2017-01-20-00-27.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/gulp-core-build",
+      "comment": "Remove several logging utilities from the public API and improve documentation in other places.",
+      "type": "minor"
```
Comment (nickpape-msft) [inline common/changes/nickpape-improve-api_2017-01-20-00-27.json]: I suppose, but these API's aren't supposed to be public in the first place

File: gulp-core-build/src/GulpProxy.ts [hunk 5]
```diff
@@ -20,7 +23,7 @@ export class GulpProxy extends Orchestrator implements gulp.Gulp {
   /* tslint:disable-next-line:no-any */
   public task(): any {
     throw new Error(
-      'You should not define gulp tasks, but instead subclass the GulpTask and override the executeTask method.'
+      'You should not define gulp tasks directly, but instead subclass the GulpTask and register it to gulp-core-build.'
```
Comment (nickpape-msft) [inline gulp-core-build/src/GulpProxy.ts]: Done

File: gulp-core-build/src/tasks/GulpTask.ts (hunk 1 above)
Comment (nickpape-msft) [inline gulp-core-build/src/tasks/GulpTask.ts]: Nope, double check the `GulpTask.execute()` function. Calling `completeCallback()` is not necessary.

File: gulp-core-build/src/tasks/GulpTask.ts (hunk 2 above)
Comment (nickpape-msft) [inline gulp-core-build/src/tasks/GulpTask.ts]: Fixed

**Pull request merged**